### PR TITLE
Feature/suppressing mechanism

### PR DESCRIPTION
**Pull request by metallmonster**
This pull request contains:
1. Suppression mechanism feature
2. ATR-297 fix
3. ATR-299 fix

### CODE REVIEW

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using Acuminator.Utilities.Common;
+using System.IO;
+
+namespace Acuminator.Vsix.Utils
+{
+	internal class SuppressionFileSystemService : ISuppressionFileSystemService
+	{
+		public XDocument Load(string path)
+		{
+			if (string.IsNullOrEmpty(path))
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @SENya1990  Ok

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+﻿using Acuminator.Utilities.DiagnosticSuppression;
+using System.Xml.Linq;
+using System;
+using Acuminator.Utilities.Common;
+using System.IO;
+
+namespace Acuminator.Vsix.Utils
+{
+	internal class SuppressionFileSystemService : ISuppressionFileSystemService
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @SENya1990 right now I don't think that we need this

File: src/Acuminator/Acuminator.Vsix/AcuminatorVSPackage.cs [hunk 3]
```diff
@@ -177,6 +181,16 @@ protected override void Initialize()
 			}
 		}
 
+		private void InitializeSuppressionManager()
+		{
+			var workspace = this.GetVSWorkspace();
+			var additionalFiles = workspace.CurrentSolution.Projects
+				.SelectMany(p => p.AdditionalDocuments)
+				.Select(d => (d.FilePath, false));
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/AcuminatorVSPackage.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Acuminator.Utilities.DiagnosticSuppression
+{
+	public class SuppressionManager
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs]: @SENya1990 good point

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+				{
+					throw new InvalidOperationException($"Suppression information for assembly {assemblyName} has been already loaded");
+				}
+
+				_fileByAssembly.Add(assemblyName, suppressionFile);
+			}
+		}
+
+		public static void Init(ISuppressionFileSystemService fileSystemService,
+			IEnumerable<(string path, bool generateSuppressionBase)> additionalFiles)
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+		{
+			var suppressionFiles = additionalFiles.Where(f => SuppressionFile.IsSuppressionFile(f.path));
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+		private const string IdAttribute = "id";
+		private const string TargetElement = "target";
+		private const string SyntaxNodeElement = "syntaxNode";
+		private const string SuppressionFileExtension = ".acuminator";
+
+		internal string AssemblyName { get; }
+
+		internal string Path { get; }
+
+		internal bool GenerateSuppressionBase { get; }
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dropsonic) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @metallmonster I know at least one common scenario where there is a need for that: if a user changes a branch without closing the Visual Studio. It reloads the whole solution, and the suppression file can differ between the branches or might be absent entirely.

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+			return Messages.Contains(message);
+		}
+
+		internal static bool IsSuppressionFile(string path)
+		{
+			return SuppressionFileExtension.Equals(System.IO.Path.GetExtension(path), StringComparison.Ordinal);
+		}
+
+		internal static SuppressionFile Load(ISuppressionFileSystemService fileSystemService,
+			(string path, bool generateSuppressionBase) loadInfo)
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
+﻿using Acuminator.Utilities.Common;
+using System;
+
+namespace Acuminator.Utilities.DiagnosticSuppression
+{
+	public readonly struct SuppressMessage
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/ISuppressionFileSystemService.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Xml.Linq;
+
+namespace Acuminator.Utilities.DiagnosticSuppression
+{
+	public interface ISuppressionFileSystemService
+	{
+		XDocument Load(string path);
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/ISuppressionFileSystemService.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+		internal string AssemblyName { get; }
+
+		internal string Path { get; }
+
+		internal bool GenerateSuppressionBase { get; }
+
+		private HashSet<SuppressMessage> Messages { get; }
+
+		public HashSet<SuppressMessage> MessagesCopy => new HashSet<SuppressMessage>(Messages);
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionFile.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+		{
+			id.ThrowOnNull(nameof(id));
+			target.ThrowOnNull(nameof(target));
+			syntaxNode.ThrowOnNull(nameof(syntaxNode));
+
+			Id = id;
+			Target = target;
+			SyntaxNode = syntaxNode;
+
+			_hashCode = $"{Id}_{Target}_{SyntaxNode}".GetHashCode();
```
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs]: @SENya1990 ok

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @SENya1990 @dropsonic ok, what do you propose to do?

[the remark this answers is not in the document]
Comment (SENya1990) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: Maybe we can catch IOException and SecurityException and then do nothing/log them somehow. Ideally there could be some fallback mode which does not use the suppression mechanism but it seems like more work to implement it.

[the remark this answers is not in the document]
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: Agree. I propose just catch exceptions for now and extend this logic later if needed.

[the remark this answers is not in the document]
Comment (SENya1990) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: I agree

File: src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs [hunk 13]
```diff
@@ -34,13 +32,20 @@ public SuppressMessage(string id, string target, string syntaxNode)
...
 			Target = target;
 			SyntaxNode = syntaxNode;
-
-			_hashCode = $"{Id}_{Target}_{SyntaxNode}".GetHashCode();
 		}
 
 		public override int GetHashCode()
 		{
-			return _hashCode;
+			int hash = 17;
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs]: @metallmonster my idea was to calculate hashcode based on this algorithm once and store it like you did before in `_hashcode` private field. I just noted that the optimisation could be more effective without the creation of concatenated string during `_hashcode` calculation in the constructor if you calculate it with this method. 
However, it's different to say something about little optimisations without benchmarks so I think it's ok if you want to leave this variant.

Comment (metallmonster) [reply src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressMessage.cs]: @SENya1990 I'll bring back `_hashCode` field

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs [hunk 14]
```diff
@@ -28,5 +40,16 @@ public string GetFileName(string path)
...
 			return Path.GetFileNameWithoutExtension(path);
 		}
+
+		public ISuppressionFileWatcherService CreateWatcher(string path)
+		{
+			var directory = Path.GetDirectoryName(path);
+			var file = Path.GetFileName(path);
+			var watcher = new FileSystemWatcher(directory, file);
+
+			watcher.NotifyFilter = NotifyFilters.LastWrite;
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @metallmonster why only specify LastWrite? I'd expect almost everythyng watched here - Security, size changes, file name changes, etc. Imho only LastAccess and perhaps CreationTime should not be watched

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs [hunk 15]
```diff
@@ -28,5 +40,16 @@ public string GetFileName(string path)
 
 			return Path.GetFileNameWithoutExtension(path);
 		}
+
+		public ISuppressionFileWatcherService CreateWatcher(string path)
```
Comment (SENya1990) [inline src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @metallmonster  CreateWatcher method is not used currently. Will it be used in the future?

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs (hunk 14 above)
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @SENya1990 good point

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs (hunk 15 above)
Comment (metallmonster) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: @SENya1990 it is used, this is probably a VS bug

**Comment by metallmonster** [comment]
@SENya1990 please review one more time

File: src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs (hunk 15 above)
Comment (SENya1990) [reply src/Acuminator/Acuminator.Vsix/Utils/SuppressionFileSystemService.cs]: ok

**Pull request merged**